Add Redis readiness verification (#3555)

ManelCoutinhoSensei opened the pull request:
### Pull Request check-list

_Please make sure to review and check all of these items:_

- [x] Do tests and lints pass with this change?
- [x] Do the CI tests pass with this change (enable it first in your forked repo and wait for the github action build to finish)?
- [x] Is the new or changed code fully tested?
- [ ] Is a documentation update included (if this change modifies existing APIs, or introduces new ones)?
- [ ] Is there an example added to the examples folder (if applicable)?
- [ ] Was the change added to CHANGES file?

_NOTE: these things are not required to open a PR and can be done
afterwards / while the PR is open._

### Description of change

Fixes #3555 (specially the [comment](https://github.com/redis/redis-py/issues/3555#issuecomment-2729673443) on that issue)

Please check if you'd rather run the tests with `check_ready=True` by default (I also tested that) and if you think that any other tests deserve a second version with this flag set.

ManelCoutinhoSensei commented:
@petyaslavova you need to rerun the failed tests due to the test suit flakiness

ManelCoutinhoSensei commented:
Hi @petyaslavova , any news about this one?

petyaslavova commented:
Hi @ManelCoutinhoSensei, I have started the review, but couldn't complete it. 
We have some urgent changes we need to complete, and the PR reviews will be delayed by a week or two. 
Sorry for the delay!

ManelCoutinhoSensei commented:
Should I solve the conflicts because the PR reviews are going to be done this week or should I postpone it?

ManelCoutinhoSensei commented:
Hey @petyaslavova! 🙌 Just checking in to see if there are any updates or an estimated timeline for moving forward with this PR. Totally understand things can get busy — just wanted to make sure it’s still on your radar.
Thanks again for your time and support! 😊

ManelCoutinhoSensei commented:
![cyring-in-the-rain](https://github.com/user-attachments/assets/6992b7f7-83f9-4353-9256-7e2ca0557fb8)

ManelCoutinhoSensei commented on redis/client.py:
```diff
@@ -203,6 +203,7 @@ def __init__(
         encoding: str = "utf-8",
         encoding_errors: str = "strict",
         decode_responses: bool = False,
+        check_ready: bool = False,
```
I'm not 100% sure what you mean by `pydocs`.
I added to the `docstrings` of both methods, let me know if that was what was intended.

ManelCoutinhoSensei commented on redis/connection.py:
```diff
@@ -378,12 +380,35 @@ def connect(self):
+        # Doing handshake since connect and send operations work even when Redis is not ready
+        if self.check_ready:
+            try:
+                ping_parts = self._command_packer.pack("PING")
```
I understand the desired but in order to use the `read_response` part of the `_send_ping`, the parsers should already be initialized (which only happens on `on_connect_check_health` that is called after the `_connect_check_server_ready`).
Either way I refactored the `_connection` method so that i could use the `send_command` command part, let me know if this suffices or if we should consider refactoring the connection flow to accommodate an earlier parser initialization

ManelCoutinhoSensei commented on redis/asyncio/connection.py:
```diff
@@ -933,7 +968,6 @@ async def _connect(self):
             reader, writer = await asyncio.open_unix_connection(path=self.path)
         self._reader = reader
         self._writer = writer
-        await self.on_connect()
```
Two reasons:
1. Make the connection flow consistent between sync and async versions
2. It wasn't present in the normal `Connection` making it inconsistent between the different async connection objects

---
This is the connection flow:
```
connect → connect_check_health:
         - _connect_check_ready (with retries) → _connect
         - on_connect_check_health(True)
-----
on_connect → on_connect_check_health(True)
```
This means that adding `on_connect` to the end of the `UnixDomainSocket._connect` will only duplicate the `on_connect_check_health`  call (but I might be missing something)

ManelCoutinhoSensei commented:
Hey @petyaslavova, just updated this PR after the merge of #3601 
One of the failing test-suits is the same problem as before and the other seems to be flaky (since it passed in my [run](https://github.com/ManelCoutinhoSensei/redis-py/actions/runs/16597016131?pr=2)).
Let me know if there are any follow ups needed to your comments or any new changes that you'd like to see made

petyaslavova commented:
> Hey @petyaslavova, have you had the chance of reviewing the changes you asked?

Hi @ManelCoutinhoSensei, I was in vacation for a while and I didn't have a chance to look at this. I'm trying to catch up with the PRs this and next week.

ManelCoutinhoSensei commented:
Hey @petyaslavova, I updated it and it is passing all the tests on my side

petyaslavova commented:
> Hey @petyaslavova, I updated it and it is passing all the tests on my side

Hi @ManelCoutinhoSensei, there are some details in this PR that can't be merged into the current master state - there is a new functionality in the client that handles specific push notifications, which can be received with the command responses. If you read directly from the socket, you won't be able to use the automatic handling of those messages, which happens in the parser. 
Can't it be transformed to use `read_response` function(the parser is already initialized when those requests are sent)?

ManelCoutinhoSensei commented:
> > > (...) Can't it be transformed to use read_response function(the parser is already initialized when those requests are sent)?
> > 
> > 
> > does [this comment](https://github.com/redis/redis-py/pull/3596#discussion_r2176235275) from July 1st address what you are asking? Or is it no longer up to date?
> 
> I have to admit I’d forgotten about this one, but yes — this is exactly the right solution here.

Which part is the right solution? Are you saying that using the `send_command` is sufficient or that we should refactor the connection flow to initialize the parser earlier?

petyaslavova commented:
> > The parser is initialized in the init method when resp3 is used - https://github.com/redis/redis-py/blob/8403ddcfcf16c95baab330076b4905483e44fbb7/redis/connection.py#L429C25-L429C37
> > This logic can be extended to set the parser when protocol is 2 as well,
> 
> I might be wrong but even when using resp3, the parser is only initialized (meaning only the `socket_read_size` is set). For the `read_response` method to work, we need to call the `Parser.on_connect` method (that initializes the `_buffer` and the `enconder`) which is only called afterwards on `on_connect_check_health` (like previously explained)

You’re absolutely right. 🙂

I finally had time to carefully review (and debug) the code related to the reported issue and this PR. My conclusion is that this change would introduce additional complexity without fully resolving the core problem.

With the current approach, we’d force a PING on all connections, even though we already have health checks available. When health checks are enabled, a PING is executed just before the first real command as part of the initial client setup.
The issue is that some clients of the library may prefer not to enable health checks because they add extra command executions.

In my opinion, the correct solution is to move the initial health-check logic (executed after the socket connect) into the retry block. That way, we can retry the entire connection-establishment flow—not just the socket connect.

Something like the following will handle the retry of the whole connection establishment logic:
`    
def connect(self):
        "Connects to the Redis server if not already connected"
        self.retry.call_with_retry(
            lambda: self.connect_check_health(
                check_health=True, retry_socket_connect=False
            ),
            lambda error: self.disconnect(error),
        )
`
This will try once the socket connect and run the commands in the `connect_check_health` method, in case of failure of any of those two - this whole flow will be retried after disconnect is called as error handler.

@ManelCoutinhoSensei, what do you think about this change in direction?

ManelCoutinhoSensei commented:
Hi again @petyaslavova 👋,

I’ve been experimenting with [this other issue](https://github.com/redis/redis-py/issues/3203) (better explained in [this comment](https://github.com/valkey-io/valkey-py/issues/169#issuecomment-2704939737)) and realized that my previous explanation wasn’t entirely accurate.

Here’s the situation:

* If the connection is successfully established but the container later goes idle, we attempt to reconnect.
* However, since the last `check_health` may have been performed very recently, that step will be skipped.

This was something I didn’t account for in my [previous comment](https://github.com/redis/redis-py/pull/3596#issuecomment-3352915216).

To address this, I updated the synchronous version with a proposal where the `PING` command is moved **unconditionally** outside of `send_commands`. In other words, regardless of the last ping or the `health_check_interval`, we always check the connection when starting a new one.

I’ve also removed the previous flag since, as mentioned earlier, the library’s default behavior already includes sending this `PING`. This change keeps that same behavior (though I can re-add the flag if you prefer).

On top of that, this approach also resolves the #3203 (mentioned before) - at least in the way I’ve been reproducing it.

Eager to hear your thoughts! Should we move forward with this solution and implement it for the async version as well? 🚀

ManelCoutinhoSensei commented:
Sorry for pinning you again, but did you have the chance to look at the idea behind this new PR @petyaslavova ? It can potentially solve 2 open PRs...
